### PR TITLE
[GOBBLIN-1530] Add error reporting for missing source nodes that may not exist

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1530

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

When copying data from a source node that does not exist/fails to be added to GaaS FlowGraph, a very cryptic error message pops up where DmasAuthorizer shows a NullPointerException.

This PR guards against source nodes that do not exist so that the error is much easier for developers and users to understand.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompiler.java [hunk 1]
```diff
@@ -190,6 +190,11 @@ public void awaitHealthy() throws InterruptedException {
     String destination = ConfigUtils.getString(flowSpec.getConfig(), ServiceConfigKeys.FLOW_DESTINATION_IDENTIFIER_KEY, "");
 
     DataNode sourceNode = this.flowGraph.getNode(source);
+    if (sourceNode == null) {
+      flowSpec.getCompilationErrors()
+          .add(String.format("Flowgraph does not have a node with id %s", source));
+      return null;
+    }
     List<String> destNodeIds = ConfigUtils.getStringList(flowSpec.getConfig(), ServiceConfigKeys.FLOW_DESTINATION_IDENTIFIER_KEY);
     List<DataNode> destNodes = destNodeIds.stream().map(this.flowGraph::getNode).collect(Collectors.toList());
```
Comment (jack-moseley) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompiler.java:199]: The same error happens if the destination node is invalid, so we should have validation on this one as well.

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompilerTest.java [hunk 2]
```diff
@@ -670,6 +670,28 @@ public void testUnresolvedFlow() throws Exception {
...
+
+    Dag<JobExecutionPlan> dag = specCompiler.compileFlow(spec);
+
+    Assert.assertEquals(dag, null);
+    Assert.assertEquals(spec.getCompilationErrors().size(), 1);
+    spec.getCompilationErrors().stream().anyMatch(s -> s.contains("Flowgraph does not have a node with id"));
+  }
+
+  @Test (dependsOnMethods = "testMissingSourceNodeError")
+  public void testMissingDestinationNodeError() throws Exception {
```
Comment (jack-moseley) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompilerTest.java:685]: Am I missing something, or shouldn't this require validation on the `List<DataNode> destNodes` object? Right now it's only validating the `DataNode sourceNode` right?

Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompilerTest.java:685]: Ah my bad I added that way too fast, good catch

Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompilerTest.java:685]: @jack-moseley you're right, it reports the error correctly on local which is what I tested, but when we add our dmas authorizer it runs into the nullpointer exception before it does another check when calculating the path, I'll add the error check now.

**Review by jack-moseley — APPROVED** [review]

**Comment by Will-Lo** [comment]
@aplex Could you merge?

**Pull request merged**